### ISSUE TITLE
Presence status for contact list?

### ISSUE THREAD

**Issue by swrobel** [yakyak/yakyak#103]
Is there any way to see online status for contacts so I know before messaging them whether they're online? Like the green dot in the hangouts chrome app:

<img width="82" alt="screen shot 2016-01-05 at 5 24 35 pm" src="https://cloud.githubusercontent.com/assets/157270/12132529/599193b0-b3d1-11e5-977b-f05d1de35491.png">

**Comment by lernerd** [comment]
I'm trying to find a solution to this, but I can't seem to find the reference for the Hangouts API to get status.
Any help?

**Comment by davibe** [comment]
@algesten can you give @lernerd some hints around hangupsjs internals ?

**Comment by lernerd** [comment]
I assume this needs to be implemented with some sort of timer that checks every N seconds that a user is online https://github.com/algesten/hangupsjs/blob/14d09b0e6954e3113828b7dc0fc06e4c273d5562/src/client.coffee#L325
Right?

**Comment by algesten** [comment]
actually I don't think there's any polling needed to do such a thing. there are two events:

https://github.com/algesten/hangupsjs#to-be-investigated

`self_presence` and `presence` that are to be investigated. mostly it would be a matter of deciphering how these packets are structured (which probably already is done) and when they are sent.

i suspect it's not much work to do this.

**Comment by lernerd** [comment]
awesomeness! I'll look into it during the week
I'm not very familiar with electron apps so it might take a while to actually implement, but I'll try my best :blush:

**Comment by lernerd** [comment]
Ok, I've played around a bit and I think the solution should be built with a combination of the call `client.queryPresence` and the event `client.on('presence'`

I propose this cuz I see that the event `presence` doesn't clearly detail the status of the user (online/offline). So after that there should be a `queryPresence` call that returns something like this:
```json
{
 "response_header": {
  "status": "OK",
  "debug_url": "",
  "request_trace_id": "14069111223457896329",
  "current_server_time": "1452605601488000"
 },
 "presence_result": [
  {
   "user_id": {
    "gaia_id": "116570108630334031607",
    "chat_id": "116570108630334031607"
   },
   "presence": {
    "reachable": true,
    "available": true,
    "device_status": {
     "mobile": false,
     "desktop": false,
     "tablet": false
    }
   }
  }
 ]
}
```
What I don't like about this approach is that at boot we should query the presence of all the contacts, right?
Any thoughts are welcomed!

**Comment by x-tk** [comment]
Presence would be really nice.  I'm used to using the buddy list that shows all of that with messages but much prefer this client overall.

**Comment by lernerd** [comment]
I still didn't have time to figure this out... I'll keep trying, any feedback is welcomed

**Comment by tdryer** [comment]
[PresenceNotification](http://hangups.readthedocs.org/en/stable/protobuf.html#presencenotification) should provide the same presence information that the query presence request does, so a separate call shouldn't be necessary. The Protocol Buffer schema is present in hangups, but it looks like hangupsjs would need an update.

**Comment by algesten** [comment]
thanks @tdryer. I need to go through the protobuf mappings to bring them in line with your original.

**Comment by davibe** [comment]
please give it some love

**Comment by kierans** [comment]
I am really missing this feature since moving to YakYak

**Comment by ariu** [comment]
I included this feature in my fork: https://github.com/ariu/yakyak/releases
Please feel free to use it, and @maxkueng , you can include it in original yak yak.

In the conversation list, if a contact was active in the last few minutes, then it has a green dot near his profile picture (thumbnails activated).

It's a green dot for now, and only for the active state, but we can improve this feature in the future: we can:
- know if the contact was active in the last few minutes
- or if he is reachable (inactive in the last minutes but has a device still connected).
- on which device: mobile, desktop, tablet.

What do you think ? :)

**Comment by algesten** [comment]
ping @davibe

**Comment by swrobel** [comment]
@ariu sadly still no working presence status for me in the beta you just released. Got this error in the inspector:

`entity.coffee:67 Uncaught TypeError: Cannot set property 'presence' of undefinedfuncs.setPresence @ entity.coffee:67(anonymous function) @ dispatcher.coffee:49doAction @ /Applications/yakyak.app/Contents/Resources/app/node_modules/trifl/lib/action.js:45action @ /Applications/yakyak.app/Contents/Resources/app/node_modules/trifl/lib/action.js:21(anonymous function) @ app.coffee:42(anonymous function) @ app.coffee:37emitOne @ events.js:77emit @ events.js:169ipcRenderer.emit @ /Applications/yakyak.app/Contents/Resources/atom.asar/renderer/api/ipc.js:16`

**Comment by randymorris** [comment]
I think this can probably be closed now.  I can say for sure it's working for me in the current prerelease but I believe this was merged just before the 1.3.0 release.
